platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,275 @@
+       withRequest: {
+         method: 'GET',
+         path: '/v0/search',
+         query: { query: 'testing' },
+         headers: {
+           'X-Key-Inflection': 'camel',
+         },
+       },
+       willRespondWith: {
+         status: 200,
+         headers: {
+           'Content-Type': 'application/json; charset=utf-8',
+         },
+         body: {
+           // The minimum expected response body.
+           data: { ... },
+           meta: { ... },
+         },
+       },
+     };
+     ```
+     - **The `state` attribute will be used as a key for the API to set up the corresponding state when verifying the contract.** For example, this might entail spinning up a mock third-party service, seeding the database, or simulating a logged in session.
+     - **The body does not need to be comprehensive.** At a minimum, the expected response only needs to define the attributes that are relevant to the test. This means that any extra, unused, or unexpected properties of the response will not break the verification later.
+   - Invoke the relevant methods that make the request being tested.
+   - Make assertions on the expected response.
+
+   **To reiterate, the tests should focus on the components or functions that make requests to the API.** It's not meaningful to simply test a fetch call to an endpoint.
+
+   There is some boilerplate in declaring the mock server and some `before`/`after`/`afterEach` hooks that perform some Pact-specific ceremony. We could simplify all of that boilerplate by writing a reusable helper function (`contractTest()`) that can be invoked like this:
+   ```
+   contractTest('VA.gov Search', 'VA.gov API', mockApi => {
+     describe('GET /search', () => {
+       context('with no results', () => {
+         it('responds with success', async () => {
+           await mockApi.addInteraction(INTERACTIONS.searchWithNoResults);
+           ...
+         })
+       })
+     })
+   });
+   ```
+2. The contract tests will run in CI when FE changes are pushed.
+   - **If they pass, they generate pacts in the form of JSON files in the `pacts` directory.** JSON files are generated for each pairing of consumer (the FE application) and provider (`vets-api`).
+   - If they fail, it's possible the request wasn't picked up by the mock server because request defined in the interaction didn't match the request that was actually made. It's also possible the component or function that made the request failed an assertion.
+3. **The CI job invokes a script to publish the pacts to a broker.** The command to publish should require a version to label the consumer and optionally accept tags. The published pacts should be versioned to the current commit hash of `vets-website`. For builds to be deployed, the pacts should also be tagged with the appropriate environment. *Tagging is important to ensure that we verify only the pact associated with the build that is getting deployed and not pacts generated from other branches.*
```

Comment:
I see this happening even before a PR is merged. It should happen as one of the build jobs.

We don't want PRs to merge that could introduce breaking changes because the `vets-api` master branch doesn't accommodate the changes yet. We also don't want to stop the `vets-website` master from deploying (to staging or production) when an app has a broken contract, because it would stop other apps from deploying their changes when they haven't broken their contracts.